compositor_pipeline/src/queue.rs

```diff
@@ -19,58 +21,100 @@ pub enum QueueError {
     output_framerate: Framerate,
+
+    /// - When new input is connected and sends the first frame we want to wait
+    /// buffer_duration before sending first frame of that input.
+    /// - When pipeline is started we want to start with a frame that was receive
+    /// `buffer_duration` time ago
     buffer_duration: Duration,
+
+    /// Base time that is used to synchronize PTS value of received frame to
+    /// the same clock. When enqueueing the frame we are modifying it's PTS
+    /// using (per input) offset calculated based on this clock.
+    ///
+    /// The end goal is that resulting PTS should be in a time frame where PTS
+    /// is equivalent to clock_start time of real time.
+    clock_start: Instant,
 }
 
 impl Queue {
     pub fn new(output_framerate: Framerate) -> Self {
         Queue {
-            internal_queue: Mutex::new(InternalQueue::new(output_framerate)),
+            internal_queue: Mutex::new(InternalQueue::new()),
             check_queue_channel: unbounded(),
             output_framerate,
             buffer_duration: DEFAULT_BUFFER_DURATION,
+            clock_start: Instant::now(),
         }
     }
 
     pub fn add_input(&self, input_id: InputId) {
-        let mut internal_queue = self.internal_queue.lock().unwrap();
-        internal_queue.add_input(input_id);
+        self.internal_queue.lock().unwrap().add_input(input_id);
     }
 
-    #[allow(dead_code)]
     pub fn remove_input(&self, input_id: &InputId) {
-        let mut internal_queue = self.internal_queue.lock().unwrap();
-        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
-        internal_queue.remove_input(input_id);
+        self.internal_queue.lock().unwrap().remove_input(input_id);
     }
 
     pub fn start(self: &Arc<Self>, sender: Sender<FrameSet<InputId>>) {
         let queue = self.clone();
         let tick_duration = self.output_framerate.get_interval_duration();
-        let buffer_duration = self.buffer_duration;
 
         QueueThread::new(
             queue,
             sender,
             queue_thread::Options {
-                buffer_duration,
                 tick_duration,
+                buffer_duration: self.buffer_duration,
+                output_framerate: self.output_framerate,
+                clock_start: self.clock_start,
             },
         )
         .spawn();
     }
 
     pub fn enqueue_frame(&self, input_id: InputId, frame: Frame) -> Result<(), QueueError> {
+        let is_first_frame = self
+            .internal_queue
+            .lock()
+            .unwrap()
+            .input_offset(&input_id)
+            .is_none();
+        if is_first_frame {
+            // Sleep here ensures that we will buffer `self.buffer_duration` on each input.
+            // It also makes calculation easier because PTS of frames will be already offset
+            // by a correct value.
+            thread::sleep(self.buffer_duration);
+        }
```

Comment:
1. At first, I understood `is_first_frame` as a first frame from all inputs (first frame enqueued).
2. I don't like to rely on inserting `input_offset` on the first frame. 

We can solve this by adding some `has_received_frame` function in the `internal queue`